etest/profile.py

```diff
@@ -0,0 +1,124 @@
+
+    pass
+
+
+class Profile:
+    """A class to generate a profile string and information.
+
+    Uses cli options.
+    """
+
+    def __init__(
+        self,
+        logger: logging.Logger,
+        strict: bool,
+        arch: str,
+        libc: str,
+        hardened: bool,
+        multilib: bool,
+        systemd: bool,
+    ) -> None:
+        """Initialize parameters."""
+        self.logger = logger
+        self.strict = strict
+
+        self.arch = arch
+        self.docker_arch = arch
+        self.pkg_arch = arch
+
+        self.libc = libc
+        self.hardened = hardened
+
+        self.multilib = multilib
+
+        self.systemd = systemd
+
+        self.logger.info("Standarizing profile data.")
+        self._standarize()
+        self.logger.info("Building profile strings.")
+        self._build()
+
+    def _warn(self, message: str) -> None:
+        """Give a warning to the user."""
+        if not self.strict:
+            self.logger.warning(message)
+        else:
+            self._error(message)
+
+    def _error(self, message: str) -> None:
+        """Raise an error."""
+        self.logger.error(message)
+        raise InvalidProfileError(message)
+
+    def _standarize(self) -> None:
+        """Sanitize profile settings."""
+        if "arm" in self.arch:
+            self.docker_arch = _arm_mappings[self.arch].value
+
+            if "armv" in self.arch:
+                self.pkg_arch = "arm"
+
+            if self.hardened or self.libc != "glibc":
+                self._error("The ARM architecture can't use a different libc or hardened profiles.")
+            if self.systemd and self.arch != "arm64":
+                self._error(f"{self.arch} doesn't support systemd.")
+
+        if self.arch == "ppc64":
+            self.docker_arch = "ppc64le"
+
+            if self.libc == "uclibc":
+                self._error("The PPC64 architecture doesn't support uclibc.")
+            if self.libc == "musl" and not self.hardened:
+                self._error("The PPC64 architecture doesn't support vanilla musl.")
+            if self.libc == "glibc" and self.hardened:
+                self._error("The PPC64 architecture doesn't support hardened glibc.")
+
+        if self.arch == "x86":
+            if self.libc == "musl" and self.hardened:
+                self._error("The x86 architecture doesn't support hardened musl.")
+
+        if self.arch != "amd64" and not self.multilib:
+            self.multilib = True
+            self._warn("--no-multilib is specific to AMD64.")
+
+        if self.libc != "glibc" and not self.multilib:
+            self._error("Alternative libcs dont support no-multilib.")
+
+        if self.systemd and (self.hardened or not self.multilib or self.libc != "glibc"):
+            self._error("Systemd profiles doesn't support alternative libcs, hardening or no multilib.")
+
+    def _build(self) -> None:
+        """Build profile information."""
+        self.profile = ""
```

Comment:
If self.profile is a class level property why do we set it here?   That would effectively wipe it when we run this command which defeats this being a dataclass looking class.

This looks like it should be a memoised property that we can create derivative properties from.  Let me know if that makes sense or if you need a bit more direction.

Reply:
I dont quite understand

Reply:
I'd suggest changing _build into a set of dynamic properties instead (see https://docs.python.org/3/library/functions.html#property).  For example:

```
@property.get
def profile(self) -> str:
    ...
```

Reply:
To memoise that you'd use functools: https://docs.python.org/3/library/functools.html#functools.cached_property